termux-shared/src/main/java/com/termux/shared/file/FileUtils.java

```diff
@@ -0,0 +1,1627 @@
+                if (permissionsToCheck != null) {
+                    // Check if permissions are missing
+                    return checkMissingFilePermissions(context, label + "regular", filePath, permissionsToCheck, false);
+                }
+            }
+        } catch (Exception e) {
+            return context.getString(R.string.error_validate_file_existence_and_permissions_failed_with_exception, label + "file", filePath, e.getMessage());
+        }
+
+        return null;
+
+    }
+
+    /**
+     * Validate the existence and permissions of directory file at path.
+     *
+     * If the {@code parentDirPath} is not {@code null}, then creation of missing directory and
+     * setting of missing permissions will only be done if {@code path} is under
+     * {@code parentDirPath} or equals {@code parentDirPath}.
+     *
+     * @param context The {@link Context} to get error string.
+     * @param label The optional label for the directory file. This can optionally be {@code null}.
+     * @param filePath The {@code path} for file to validate or create. Symlinks will not be followed.
+     * @param parentDirPath The optional {@code parent directory path} to restrict operations to.
+     *                      This can optionally be {@code null}. It is not canonicalized and only normalized.
+     * @param createDirectoryIfMissing The {@code boolean} that decides if directory file
+     *                                 should be created if its missing.
+     * @param permissionsToCheck The 3 character string that contains the "r", "w", "x" or "-" in-order.
+     * @param setPermissions The {@code boolean} that decides if permissions are to be
+     *                              automatically set defined by {@code permissionsToCheck}.
+     * @param setMissingPermissionsOnly The {@code boolean} that decides if only missing permissions
+     *                                  are to be set or if they should be overridden.
+     * @param ignoreErrorsIfPathIsInParentDirPath The {@code boolean} that decides if existence
+     *                                  and permission errors are to be ignored if path is
+     *                                  in {@code parentDirPath}.
+     * @param ignoreIfNotExecutable The {@code boolean} that decides if missing executable permission
+     *                              error is to be ignored. This allows making an attempt to set
+     *                              executable permissions, but ignoring if it fails.
+     * @return Returns the {@code errmsg} if path is not a directory file, failed to create it,
+     * or validating permissions failed, otherwise {@code null}.
+     */
+    public static String validateDirectoryFileExistenceAndPermissions(@NonNull final Context context, String label, final String filePath, final String parentDirPath, final boolean createDirectoryIfMissing,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 25. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.